### PR TITLE
Quickicon improvements

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #258  .

### Summary of Changes
1) Fixed: Button MenuItems
 
2) Numbers on icons can be useful and important but sometimes are useless, asdescribe in the issues.
We have expanded module parameters so the administrator can choose if she want to see the number or only the button.

3) If a number is displayed, both the number and component name have attribute "aria-hidden", but a sr-only text is added which gives information in a sentence.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 1]
```diff
@@ -8,33 +8,50 @@ MOD_QUICKICON_ADD_NEW="New %s"
 MOD_QUICKICON_AMOUNT_THSD="THSD"
 MOD_QUICKICON_ARTICLE_MANAGER="Articles"
 MOD_QUICKICON_ARTICLE_MANAGER_1="Article"
-MOD_QUICKICON_CACHE="Cache (MB)"
+MOD_QUICKICON_ARTICLE_SRONLY="Articles: %s articles are available"
+MOD_QUICKICON_ARTICLE_SRONLY_0="Articles: No  is available"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini]:
```suggestion
MOD_QUICKICON_ARTICLE_SRONLY_0="Articles: No article is available"
```

??

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 2]
```diff
@@ -8,33 +8,50 @@ MOD_QUICKICON_ADD_NEW="New %s"
...
 MOD_QUICKICON_INSTALL_EXTENSIONS="Install Extensions"
-MOD_QUICKICON_LANGUAGE_MANAGER="Languages"
+MOD_QUICKICON_LOAD_PLUGINS_LABEL="Load 3rd party plugins"
 MOD_QUICKICON_MAINTENANCE="Maintenance"
 MOD_QUICKICON_MEDIA_MANAGER="Media"
 MOD_QUICKICON_MENUITEMS_ADD="New Menu Item"
 MOD_QUICKICON_MENUITEMS_MANAGER="Menu Items"
 MOD_QUICKICON_MENUITEMS_MANAGER_1="Menu Item"
+MOD_QUICKICON_MENUITEMS_SRONLY="Menu Items. %s Menu Items are available."
+MOD_QUICKICON_MENUITEMS_SRONLY_0="Menu Items. No Menu Items are available."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini]:
```suggestion
MOD_QUICKICON_MENUITEMS_SRONLY_0="Menu Items. No Menu Items is available."
```

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 3]
```diff
@@ -63,4 +79,7 @@ MOD_QUICKICON_UPDATE="Icons Update"
 MOD_QUICKICON_USERS="Users"
 MOD_QUICKICON_USER_MANAGER="Users"
 MOD_QUICKICON_USER_MANAGER_1="User"
+MOD_QUICKICON_USER_SRONLY="Users: %s Users are activated."
+MOD_QUICKICON_USER_SRONLY_1="Users: One User is activated."
+MOD_QUICKICON_WITH_COUNT="Show and count Items"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini]:
```suggestion
MOD_QUICKICON_WITH_COUNT="Show Items with Count"
```

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 4]
```diff
@@ -8,33 +8,50 @@ MOD_QUICKICON_ADD_NEW="New %s"
...
 MOD_QUICKICON_MODULE_ADD="New Module"
 MOD_QUICKICON_MODULE_MANAGER="Modules"
 MOD_QUICKICON_MODULE_MANAGER_1="Module"
+MOD_QUICKICON_MODULE_SRONLY="Modules. %s Modules are available"
+MOD_QUICKICON_MODULE_SRONLY_0="Modules. No Module is available"
+MOD_QUICKICON_MODULE_SRONLY_1="Modules: One Module is available"
 MOD_QUICKICON_NAV_LABEL="Quick Links"
 MOD_QUICKICON_PLUGIN_MANAGER="Plugins"
+MOD_QUICKICON_PLUGIN_SRONLY="Plugins. %s Plugins are available"
+MOD_QUICKICON_PLUGIN_SRONLY_1="Modules: One Plugin is available"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini]:
```suggestion
MOD_QUICKICON_PLUGIN_SRONLY_1="Plugins: One Plugin is available"
```

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 5]
```diff
@@ -8,33 +8,50 @@ MOD_QUICKICON_ADD_NEW="New %s"
...
 MOD_QUICKICON_CONTENT="Content"
-MOD_QUICKICON_EXTENSIONS="Extensions"
-MOD_QUICKICON_EXTENSION_MANAGER="Extension Manager"
 MOD_QUICKICON_GLOBAL_CONFIGURATION="Global Configuration"
 MOD_QUICKICON_GROUP_DESC="The group of this module (this value is compared with the group value used in <strong>Quick Icons</strong> plugins to inject icons). The 'mod_quickicon' group always displays the Joomla! core icons."
 MOD_QUICKICON_GROUP_LABEL="Group"
 MOD_QUICKICON_HEADER_ICON_LABEL="Header Icon"
 MOD_QUICKICON_INSTALL_EXTENSIONS="Install Extensions"
-MOD_QUICKICON_LANGUAGE_MANAGER="Languages"
+MOD_QUICKICON_LOAD_PLUGINS_LABEL="Load 3rd party plugins"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini:23]: plugins? do you mean extensions? Is this even used?

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 6]
```diff
@@ -8,33 +8,50 @@ MOD_QUICKICON_ADD_NEW="New %s"
 MOD_QUICKICON_AMOUNT_THSD="THSD"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini]: What is this? Is it used?

File: administrator/modules/mod_quickicon/mod_quickicon.xml [hunk 7]
```diff
@@ -40,8 +40,8 @@
 				<field
 					name="show_global"
 					type="radio"
-					label="MOD_QUICKICON_SHOW_GLOBAL_LABEL"
 					class="switcher"
+					label="MOD_QUICKICON_SHOW_GLOBAL_LABEL"
```
Comment (brianteeman) [inline administrator/modules/mod_quickicon/mod_quickicon.xml]: Please put back in the original order as the codestyle is name, type, label

File: administrator/modules/mod_quickicon/mod_quickicon.xml [hunk 8]
```diff
@@ -50,8 +50,8 @@
 				<field
 					name="show_checkin"
 					type="radio"
-					label="MOD_QUICKICON_SHOW_CHECKIN_LABEL"
 					class="switcher"
+					label="MOD_QUICKICON_SHOW_CHECKIN_LABEL"
```
Comment (brianteeman) [inline administrator/modules/mod_quickicon/mod_quickicon.xml]: Please put back in the original order as the codestyle is name, type, label

File: administrator/modules/mod_quickicon/mod_quickicon.xml [hunk 9]
```diff
@@ -60,88 +60,88 @@
 				<field
 					name="show_cache"
 					type="radio"
-					label="MOD_QUICKICON_SHOW_CACHE_LABEL"
 					class="switcher"
+					label="MOD_QUICKICON_SHOW_CACHE_LABEL"
```
Comment (brianteeman) [inline administrator/modules/mod_quickicon/mod_quickicon.xml]: Please put back in the original order as the codestyle is name, type, label

File: administrator/modules/mod_quickicon/mod_quickicon.xml [hunk 10]
```diff
@@ -60,88 +60,88 @@
...
 					<option value="1">JSHOW</option>
+					<option value="2">MOD_QUICKICON_WITH_COUNT</option>
 				</field>
 				<field
-					name="show_menuItems"
+					name="show_media"
 					type="radio"
-					label="MOD_QUICKICON_SHOW_MENUITEMS_LABEL"
 					class="switcher"
+					label="MOD_QUICKICON_SHOW_MEDIA_LABEL"
```
Comment (brianteeman) [inline administrator/modules/mod_quickicon/mod_quickicon.xml:113]: Please put back in the original order as the codestyle is name, type, label

File: administrator/modules/mod_quickicon/mod_quickicon.xml [hunk 11]
```diff
@@ -60,88 +60,88 @@
...
 					<option value="0">JHIDE</option>
 					<option value="1">JSHOW</option>
+					<option value="2">MOD_QUICKICON_WITH_COUNT</option>
 				</field>
 				<field
 					name="show_templates"
 					type="radio"
-					label="MOD_QUICKICON_SHOW_TEMPLATES_LABEL"
 					class="switcher"
+					label="MOD_QUICKICON_SHOW_TEMPLATES_LABEL"
```
Comment (brianteeman) [inline administrator/modules/mod_quickicon/mod_quickicon.xml]: Please put back in the original order as the codestyle is name, type, label

**Comment by brianteeman** [comment]
Made some comments inline

Someone else will have to check that the pluralisation is correct in the code to allow for languages that have more complex pluralisation rules - maybe @infograf768 can check this

File: administrator/language/en-GB/en-GB.mod_quickicon.ini (hunk 5 above)
Comment (chmst) [reply administrator/language/en-GB/en-GB.mod_quickicon.ini:23]: We don't have icons for languages and extensions. So better delete the language keys

File: administrator/language/en-GB/en-GB.mod_quickicon.ini (hunk 6 above)
Comment (chmst) [reply administrator/language/en-GB/en-GB.mod_quickicon.ini]: It is outdated, thanks for the hint.
The taget was to truncate numbers > 100000 but is not longer necessary

File: administrator/components/com_cache/Controller/DisplayController.php [hunk 12]
```diff
@@ -55,7 +53,13 @@ public function getQuickiconContent()
 		}
 
 		// Number bytes are returned in format xxx.xx MB
-		$result = strstr(HTMLHelper::_('number.bytes', $size, 'MB', 1, false), '.', true);
+		$bytes = HTMLHelper::_('number.bytes', $size, 'MB', 1);
+		$result = strstr($bytes, '.', true);
+
+		// No dot found, so just take the converted size
+		if (!$result) {
```
Comment (Quy) [inline administrator/components/com_cache/Controller/DisplayController.php]:
```suggestion
		if (!$result)
		{
```

File: administrator/modules/mod_quickicon/Helper/QuickIconHelper.php [hunk 13]
```diff
@@ -65,54 +65,82 @@ public static function &getButtons(Registry $params, CMSApplication $application
...
-				self::$buttons[$key][] = [
-					'ajaxurl' => 'index.php?option=com_users&amp;task=users.getQuickiconContent&amp;format=json',
-					'image'   => 'fa fa-users',
-					'link'    => Route::_('index.php?option=com_users'),
-					'linkadd' => Route::_('index.php?option=com_users&task=user.add'),
-					'name'    => 'MOD_QUICKICON_USER_MANAGER',
-					'access'  => array('core.manage', 'com_users', 'core.create', 'com_users'),
-					'group'   => 'MOD_QUICKICON_SITE',
+				$tmp = [
+					'image'  => 'fa fa-users',
```
Comment (Quy) [inline administrator/modules/mod_quickicon/Helper/QuickIconHelper.php]: Line up `=>`

File: administrator/modules/mod_quickicon/Helper/QuickIconHelper.php [hunk 14]
```diff
@@ -128,27 +156,41 @@ public static function &getButtons(Registry $params, CMSApplication $application
...
-				self::$buttons[$key][] = [
-					'ajaxurl' => 'index.php?option=com_plugins&amp;task=plugins.getQuickiconContent&amp;format=json',
-					'image'   => 'fa fa-plug',
-					'link'    => Route::_('index.php?option=com_plugins'),
-					'name'    => 'MOD_QUICKICON_PLUGIN_MANAGER',
-					'access'  => array('core.manage', 'com_plugins'),
-					'group'   => 'MOD_QUICKICON_SITE'
+				$tmp = [
+					'image'  => 'fa fa-plug',
+					'srOnly'  => 'MOD_QUICKICON_PLUGIN_SRONLY',
```
Comment (Quy) [inline administrator/modules/mod_quickicon/Helper/QuickIconHelper.php]:
```suggestion
					'srOnly' => 'MOD_QUICKICON_PLUGIN_SRONLY',
```

File: layouts/joomla/quickicons/icon.php [hunk 15]
```diff
@@ -50,6 +53,12 @@
 // Make the class string
 $class = !empty($tmp) ? 'class="' . implode(' ', array_unique($tmp)) . '"' : '';
 
+if (isset($displayData['srOnly']))
+{
+	$srOnly_0 = Text::plural($displayData['srOnly'], 0);
+	$srOnly_1 = Text::plural($displayData['srOnly'], 1);
+	$srOnly_n = Text::plural($displayData['srOnly'], 2);
+}
 if (isset($displayData['name']))
```
Comment (Quy) [inline layouts/joomla/quickicons/icon.php]:
```suggestion

if (isset($displayData['name']))
```

File: layouts/joomla/quickicons/icon.php [hunk 16]
```diff
@@ -70,13 +79,20 @@
 			</div>
 		<?php endif; ?>
 		<?php if (isset($displayData['ajaxurl'])) : ?>
-			<div class="quickicon-amount" <?php echo $dataUrl ?>>
-				<span class="fa fa-spinner" aria-hidden="true"></span>
+			<div class="quickicon-amount" <?php echo $dataUrl ?> aria-hidden="true">
+				<span class="fa fa-spinner"></span>
 			</div>
+			<?php if (isset($displayData['srOnly']) ): ?>
```
Comment (Quy) [inline layouts/joomla/quickicons/icon.php]:
```suggestion
			<?php if (isset($displayData['srOnly']) ) : ?>
```

File: layouts/joomla/quickicons/icon.php [hunk 17]
```diff
@@ -70,13 +79,20 @@
...
+				<div class="quickicon-sr-desc sr-only"
+					data-sronly-zero="<?php echo $srOnly_0 ?>"
+					data-sronly-one="<?php echo $srOnly_1 ?>"
+					data-sronly-n="<?php echo $srOnly_n ?>">
+				</div>
+			<?php endif; ?>
 		<?php endif; ?>
 		<?php // Name indicates the component
 		if (isset($displayData['name'])): ?>
```
Comment (Quy) [inline layouts/joomla/quickicons/icon.php:79]:
```suggestion
		if (isset($displayData['name'])) : ?>
```

**Comment by infograf768** [comment]
Would someone be kind enough to explain what we are supposed to get (screenshots) and what the plural forms are for and how/where they display?

I have installed the `quickicon-improvements` branch here

**Comment by infograf768** [comment]
Note: I have enabled all possible stuff in the Quickicon module

**Comment by infograf768** [comment]
I get this when enabled
```
<div class="quickicon-sr-desc sr-only" data-sronly-zero="Article categories: No category is available" data-sronly-one="Article categories: One category is available" data-sronly-n="Article categories: 2 categories are available">Article categories: 5 categories are available</div>
```
I am lost as whatever that means it seems that any other plural form is not taken care of.
```
if (isset($displayData['srOnly']))
{
	$srOnly_0 = Text::plural($displayData['srOnly'], 0);
	$srOnly_1 = Text::plural($displayData['srOnly'], 1);
	$srOnly_n = Text::plural($displayData['srOnly'], 2);
}
```

**Comment by infograf768** [comment]
Example
In en-GB we have
```
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_1="%d banner checked in."
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_MORE="%d banners checked in."
```

in Russian
```
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_0="Ни один баннер не был разблокирован"
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_1="%d баннер успешно разблокирован"
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_2="%d баннера успешно разблокировано"
COM_BANNERS_BANNERS_N_ITEMS_CHECKED_IN_MORE="%d баннеров успешно разблокировано"
```
In some other languages, we may have even more strings/plural forms.
In Scottish (only available for 2.5) we have
```
public static function getPluralSuffixes($count) {
		if ($count == 0 || $count > 19) {
            $return =  array('0');
        }
        elseif($count == 1 || $count == 11) {
               $return =  array('1');
        }
        elseif($count == 2 || $count == 12) {
               $return =  array('2');
	    }
	    elseif(($count > 2 && $count < 12) || ($count > 12 && $count < 19)) {
                $return =  array('FEW');
		}
		return $return;
     }
```

**Comment by brianteeman** [comment]
That's exactly what I suspected

**Comment by chmst** [comment]
The Plural forms do this: 
1) for sighted users

quickicons-singular


2) For blind users the Screen Reader says for example: 

Users: One User is activated.

Articles: No article is available

Modules: Three modules are available

Global Checkin: No Items are locked.

Cache: The Cache is empty

We have adopted this from the joomla update plugins. In case of many different plural forms we could add more strings.

**Comment by brianteeman** [comment]
> In case of many different plural forms we could add more strings.

That's the whole point - we can not. The current system used throughout joomla is that the TT add them if they need to

**Comment by bembelimen** [comment]
@infograf768 you're right, we have to take care for the pluralism form. Do you have an JS example where this happen in Joomla!?

**Comment by infograf768** [comment]
Did not find any but this should help https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Language/Text.php#L171-L198

**Comment by infograf768** [comment]
@chmst 
Please use the normalized string constants for plural forms. 
I mean adding the `_N_` in the constant. For example:
`COM_CATEGORIES_QUICKICON_N_SRONLY`

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 18]
```diff
@@ -108,6 +108,12 @@ COM_MENUS_ITEM_FIELD_TITLE_LABEL="Menu Title"
 COM_MENUS_ITEM_FIELD_TYPE_LABEL="Menu Item Type"
 COM_MENUS_ITEM_MODULE_ASSIGNMENT="Module Assignment"
 COM_MENUS_ITEM_ROOT="Menu Item Root"
+COM_MENUS_ITEMS_QUICKICON="Menus Items"
+COM_MENUS_ITEMS_QUICKICON_0="Menus Items"
+COM_MENUS_ITEMS_QUICKICON_1="Menu Item"
+COM_MENUS_ITEMS_QUICKICON_SRONLY="Menu Items. %s Menu Items are available."
+COM_MENUS_ITEMS_QUICKICON_SRONLY_0="Menu Items. No Menu Items are available."
```
Comment (infograf768) [inline administrator/language/en-GB/en-GB.com_menus.ini]: Please normalise the use of `:`

File: administrator/components/com_cache/Controller/DisplayController.php [hunk 19]
```diff
@@ -55,7 +53,19 @@ public function getQuickiconContent()
...
+		// No dot found, so just take the converted size
+		if (!$mb)
+		{
+			$mb = $bytes;
+		}
+
+		$result = [];
+
+		$result['amount'] = $mb;
+		$result['sronly'] = Text::sprintf('COM_CACHE_QUICKICON_SRONLY', $mb);
```
Comment (infograf768) [inline administrator/components/com_cache/Controller/DisplayController.php]: Think RTL:
`$result['amount'] = '&#x200E;' . $mb;`
To test with (once modified to add the `_N_`)
`$result['sronly'] = Text::sprintf('COM_CACHE_QUICKICON_N_SRONLY', '&#x200E;' . $mb);`
This should display normally a LTR for the value, i.e. always `0.1 MB` and not `MB 0.1`

File: administrator/language/en-GB/en-GB.mod_quickicon.ini [hunk 20]
```diff
@@ -5,36 +5,33 @@
...
 MOD_QUICKICON_MAINTENANCE="Maintenance"
 MOD_QUICKICON_MEDIA_MANAGER="Media"
 MOD_QUICKICON_MENUITEMS_ADD="New Menu Item"
-MOD_QUICKICON_MENUITEMS_MANAGER="Menu Items"
 MOD_QUICKICON_MENUITEMS_MANAGER_1="Menu Item"
+MOD_QUICKICON_MENUITEMS_MANAGER_MORE="Menu Items"
 MOD_QUICKICON_MODULE_ADD="New Module"
-MOD_QUICKICON_MODULE_MANAGER="Modules"
+MOD_QUICKICON_MODULE_MANAGER_MORE="Modules"
 MOD_QUICKICON_MODULE_MANAGER_1="Module"
```
Comment (infograf768) [inline administrator/language/en-GB/en-GB.mod_quickicon.ini:31]: Alpha order
_1 comes before _MORE

**Comment by chmst** [comment]
Thank ou, @infograf768. 
I did the normalization but was not successful with the RTL. Will try again when I have some spare time.

**Comment by infograf768** [comment]
RTL: will test a new install of your branch and see what I can do.

File: administrator/components/com_cache/Controller/DisplayController.php [hunk 21]
```diff
@@ -65,7 +66,7 @@ public function getQuickiconContent()
 		$result = [];
 
 		$result['amount'] = $mb;
-		$result['sronly'] = Text::sprintf('COM_CACHE_QUICKICON_SRONLY', $mb);
+		$result['sronly'] = Text::sprintf('COM_CACHE_N_QUICKICON_SRONLY', $mb);
```
Comment (infograf768) [inline administrator/components/com_cache/Controller/DisplayController.php]: You do not need the `_N_` here as it not a plural form.
Therefore it also has to be modified in the Cache DisplayController.

**Comment by infograf768** [comment]
RTL: the problem with the Cache value display is the JsonResponse.
It does not accept htmlentities in general and even if it was, this specific one should not be decoded anyway as it is a pure unicode `&#x200E;` = `\u200E`

For example if I use
`$result['amount'] = json_encode('&#x200E;' . $mb);`
I get the correct LTR but with the htmlentity
Screen Shot 2019-07-13 at 19 48 11

File: administrator/components/com_cache/Controller/DisplayController.php (hunk 21 above)
Comment (infograf768) [reply administrator/components/com_cache/Controller/DisplayController.php]: Also:
I suggest to modify to not use `$mb` but only `$bytes`, as we would not get `MB` when we have plain numbers, i.e. 1, 2, etc.
Suggestion
```
		// Number bytes are returned in format xxx.xx MB
		$bytes = HTMLHelper::_('number.bytes', $size, 'MB', 1);

		$result = [];

		$result['amount'] = $bytes;
		$result['sronly'] = Text::sprintf('COM_CACHE_QUICKICON_SRONLY', $bytes);
```

**Comment by infograf768** [comment]
@chmst
 RTL: found a solution in quickicon.es6.js
Change
`counter.textContent =  response.data.amount;`
to
`counter.textContent =  '\u200E' + response.data.amount;`

Screen Shot 2019-07-14 at 12 07 45

Same for sronly
`sronly.textContent = '\u200E' + response.data.sronly;`

provided that RTL TT creates a string formatted such:
`COM_CACHE_QUICKICON_SRONLY= "%s حجم کش است"`

Screen Shot 2019-07-14 at 12 25 45

File: administrator/components/com_cache/Controller/DisplayController.php (hunk 21 above)
Comment (chmst) [reply administrator/components/com_cache/Controller/DisplayController.php]: Resolved wit a new commit with better sronly message.

**Review by bembelimen — APPROVED** [review]

**Pull request merged**